File: problems/509-fibonacci-number/main.py
class Solution:
    def fib(self, N: int) -> int:
        a = 0
        b = 1
        if (N == 0):
          return a
        if (N == 1):
          return b
        
        for _ in range(2, N + 1):
          value = a + b
          
          a = b
          b = value
          
        return value

File: problems/509-fibonacci-number/main.cpp
class Solution {
public:
  int fibonacci(int n, vector<int>& dp) {
    if (n <= 1)
      return n;
    if (dp[n] != -1)
      return dp[n];
    dp[n] = fibonacci(n - 1, dp) + fibonacci(n - 2, dp);
    return dp[n];
  }
  int fib(int N) {
    vector<int> dp(N + 1, -1);
    return fibonacci(N, dp);
  }
};
